### ISSUE TITLE
Add entrypoint for compile progress notifications

### ISSUE THREAD

**Issue by jvican** [scalacenter/bsp#7]
Build tools have ways to hook into compilers to get progress notifications. For example, this is Zinc's interface:
  
```scala
/*
 * Zinc - The incremental compiler for Scala.
 * Copyright 2011 - 2017, Lightbend, Inc.
 * Copyright 2008 - 2010, Mark Harrah
 * This software is released under the terms written in LICENSE.
 */

package xsbti.compile;

/**
 * An API for reporting when files are being compiled.
 *
 * This design is tied to the Scala compiler but it is used
 * by the Java compiler too, check the docs of the methods.
 */
public interface CompileProgress {

	/**
	 * Start the progress of a concrete phase for the path of a given
	 * compilation unit.
	 *
	 * @param phase The phase of the compiler being run.
	 * @param unitPath The path of the compilation unit. It will be empty
	 *                 when a Java compiler is reporting the progress.
	 */
	void startUnit(String phase, String unitPath);

	/**
	 * Advance the progress of the current unit.
	 *
	 * @param current The current progress.
	 * @param total The total of the progress that has to be achieved.
	 *
	 * @return Whether the user has cancelled compilation or not.
	 */
	boolean advance(int current, int total);
}
```
  
Progress notifications are useful to notify users of IDEs and editors how is compilation going, and have a feeling of what are the slow phases (or even the slow compilation units) to compile. The best way to know that you have a slow project to compile is to expose this information to the users!
  
This ticket registers my intention of adding such an entrypoint in BSP, under the name of `buildTarget/showCompileProgress`. I haven't yet figured out the right interface, but it would look similar to Zinc's interface. We may even need to define two events to support every method, and then an event kind to the `ShowCompileProgressParams`.
  
This entrypoint would allow IntelliJ, for example, to mimic the same user experience than the built-in "Build Project" which shows compilation progress notifications in a small event window at the bottom right.

**Comment by jastice** [comment]
I think the parameters need to have some kind of threading id to model the tree structure. Something like:
```
startUnit(String parentId, String id, String phase, String unitPath)
advance(String id, int current, int total)
```

**Comment by jastice** [comment]
For reference, the progress API in the IntelliJ build tool window looks like this:
```java
public StartEventImpl(
    @NotNull Object eventId, 
    @Nullable Object parentId, 
    long eventTime, 
    @NotNull String message)

public StartEventImpl(
    @NotNull Object eventId, 
    @Nullable Object parentId, 
    long eventTime, 
    @NotNull String message)

public FinishEventImpl(
    @NotNull Object eventId,
    @Nullable Object parentId,
    long eventTime,
    @NotNull String message,
    @NotNull EventResult result)
```

**Comment by jastice** [comment]
We don't have only compiles, but also tests, and potentially other types of tasks. Any of these tasks may have subtasks that the build tool may want to report in a more fine-grained manner.

For example in compile:
* started compile target
* compiled n of N files
* finished compiling

in test:
* started test suite S
* started test group G in suite S
* started test T in group G
* finished test T, group G, suite S

in arbitraty task:
* started task T
* spawned subtasks S1, S2, S3
* finished subtask S1
* canceled subtask S2, S3
* finished task T

etc.

I think this can be addressed with a common notification interface with additional messages for specialized tasks as compile and test. Request for comment:


Generic notifications:
```
build/taskStart {
    taskId: TaskId
    eventTime: Long
    message: String
}

build/taskProgress {
    taskId: TaskId
    eventTime: Long
    message: String
    total: Option[Long]
    progress: Option[Long]
    unit: String
}

build/taskFinish {
    taskId: TaskId
    eventTime: Long
    message: String
    result: StatusCode
}
```
Specific notifications:

```
// complement to buildTarget/compileReport
buildTarget/compileTask {
    taskId: TaskId
    target: BuildTargetIdentifier
}

// complement to buildTarget/testTask
buildTarget/testTask {
    taskId: TaskId
    target: BuildTargetIdentifier
}
```

Server sends specific notification first, then the generic notification with the same task id.

Example (simplified):

* client requests compile of target A
* server: `buildTarget/compileTask (taskId=1, target=A)`
* server: `build/taskStart(taskId=1, eventTime=100, message="started compiling A")`
* server: `build/taskProgress(taskId=1, eventTime=102, message="compiling A/stuff", total=Some(10), progress=Some(3))`
* server: `build/taskFinish(taskId=1, eventTime=110, message="compiling A failed")`

The drawback here is that the client would have to keep track of task "description" and task updates by combining based on `taskId`

**Comment by olafurpg** [comment]
Modeling after the IntelliJ build tool window is a great idea, your proposal for the generic notifications looks good to me @jastice . I am not sure about the `buildTarget/testTask` and `buildTarget/compileTask` notifications however, what do you think about embedding that data instead in `taskProgress` with a `data: Json` (modeled as `Object data` in bsp4j)? That would allow the server to include an additional field `data: {buildTarget:"$URI"}`.

**Comment by jastice** [comment]
Yes that seems better. We can define standard data objects that will be sent for compile, test and other builtins, but allow the server to define their own.

In that vein, should we replace compileResult / testResult notifications with taskFinish notifications with  result data object?

**Comment by olafurpg** [comment]
I am not sure if we buy much by removing `CompileReport` and `TestReport`, the two data structures are quite specific and important enough to merit their own endpoints in my opinion. I already rely on `CompileReport` in metals to push diagnostics to the client.

If we go that route and encode them in `taskProgress.data: Json` then we need to include a tag/ID to distinguish between `CompileReport` and `TestReport` since both have `target: BuildTargetIdentifier` and `originId: Option[String]`.

**Comment by olafurpg** [comment]
> In that vein, should we replace compileResult / testResult notifications with taskFinish notifications with  result data object?

After more thinking, I'm fine with this route as long as 

- there is a way to distinguish between `CompileReport` and `TestReport`.
- we decide what to do with `CompileResult`/`RunResult`/`TestResult` that currently return `statusCode: Int`. Do we remove `statusCode`?

**Comment by jastice** [comment]
* `CompileReport` and `TestReport` are distinguished by their structure.
* `CompileResult` etc are the response object to the request and should be independent from notifications, which a client can choose not to handle. I think they should stay as they are.

**Comment by jastice** [comment]
Clarification: My proposal wasn't to outright remove `CompileReport`, but embed it as data into a `taskFinish` notification.

**Comment by olafurpg** [comment]
Makes sense, I am OK with consolidating task progress notifications under your proposal 👍

**Comment by jvican** [comment]
The major issue I see with this generic task progress API is that it's redundant with `CompileReport` and `TestReport` up until the point where it doesn't make sense to have them both. Either we remove the reports altogether or we keep them as they are.
  
Is it possible that we take a simpler approach to this problem and we add ad-hoc compilation progress and test status APIs? For example, we want the build server to send a test status notification to the client every time there is a status change in either a test case or a test suite (the build server picks the granularity at which it operates). Can we not just add a `TestStatus` notification akin to the one that started this ticket?
  
If we go this route, then we're giving up this use case:
> Allowing a client to receive progress notifications about other generic tasks that are run in the build tool side. For example, running compile on sbt may trigger resolution too, and I can see a lot of value in the IDE collecting progress notifications about these generic tasks so that a final breakdown can be shown to IDE users.
  
But it's not clear to me we need it in this iteration of BSP. Maybe settling for a simpler solution will give us more time to rethink more whether a generic API is necessary after all, or even if the above use case is worth supporting (I'm not sure it is).
  
For be, the biggest deal breaker is that the generic interface proposed here requires the client to look for the `data: Json` field of the `taskFinish` notifications and then infer which report it is (whether a compilation report, test report or something else). I prefer a more explicit API.
  
What do you guys think?

**Comment by olafurpg** [comment]
> For be, the biggest deal breaker is that the generic interface proposed here requires the client to look for the data: Json field of the taskFinish notifications and then infer which report it is (whether a compilation report, test report or something else). I prefer a more explicit API.

I agree.


> I like the proposed design to do progress notifications, but I'm unsure we should push it further and try to abstract over results like CompileResult, RunResult and TestResult.

With minor additions to `Task{Start,Progress,End}` we can have first-class encodings for test/run/compile reports while still keeping the API generic

```diff
trait TaskStartParams {
     /** Optional metadata about the task.
-      * For compilation tasks, this should be a CompileTask object.
-      * For test tasks, this should be a TestTask object. */
    def data: Option[Json]
+   def testTask: Option[TestTask]
+   def compileTask: Option[CompileTask]
}
```

**Comment by jastice** [comment]
Resolved via #49
